bootloader/lkenv/lkenv.go

```diff
@@ -0,0 +1,278 @@
+type Env struct {
+	path string
+	pathbak string
+	env SnapBootSelect
+}
+
+//helper function to trim string from byte array to actual length
+func cToGoString(c []byte) string {
+	n := -1
+	for i, b := range c {
+		if b == 0 {
+			break
+		}
+		n = i
+	}
+	return string(c[:n+1])
+}
+
+// helper function to copy string into array and making sure it's terminated
+func copyString(b []byte, s string) {
+	if (s == "") {
+		b[0] = 0
+	} else {
+		copy(b[:], s)
+		b[len(s)] = 0
+	}
+}
+
+func NewEnv(path string) *Env {
+	// osutil.FileExists(path + "bak")
+	return &Env{
+		path: path,
+		pathbak: path + "bak",
+	}
+}
+
+func (l *Env) Get(key string) string {
+	switch key {
+	case "snap_mode":
+		return cToGoString(l.env.Snap_mode[:])
+	case "snap_kernel":
+		return cToGoString(l.env.Snap_kernel[:])
+	case "snap_try_kernel":
+		return cToGoString(l.env.Snap_try_kernel[:])
+	case "snap_core":
+		return cToGoString(l.env.Snap_core[:])
+	case "snap_try_core":
+		return cToGoString(l.env.Snap_try_core[:])
+	}
+	return ""
+}
+
+func (l *Env) Set(key, value string) {
+	switch key {
+	case "snap_mode":
+		copyString(l.env.Snap_mode[:], value)
+	case "snap_kernel":
+		copyString(l.env.Snap_kernel[:], value)
+	case "snap_try_kernel":
+		copyString(l.env.Snap_try_kernel[:], value)
+	case "snap_core":
+		copyString(l.env.Snap_core[:], value)
+	case "snap_try_core":
+		copyString(l.env.Snap_try_core[:], value)
+	}
+}
+
+// Configure partition labels for used boot partitions
+// this should not be used at run time, it should be used
+// once at image build time, if part labels are not filled by gadget build
+func (l *Env) ConfigureBootPartitions(boot_1, boot_2 string) {
+	copyString(l.env.Bootimg_matrix[0][0][:], boot_1)
+	copyString(l.env.Bootimg_matrix[1][0][:], boot_2)
+}
+
+func (l *Env) Load() error {
+	err := l.LoadEnv(l.path)
+	if (err != nil) {
+		return l.LoadEnv(l.pathbak)
+	}
+	return nil
+}
+
+func (l *Env) LoadEnv(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("error loading lk environment file: %v", err)
+	}
+
+	defer f.Close()
+	if err := binary.Read(f, binary.LittleEndian, &l.env); err != nil {
+		return fmt.Errorf("error reading lk environment from file %v", err)
+	}
+	// calculate crc32 to validate structure
+	w := bytes.NewBuffer(nil)
+	ss := binary.Size(l.env)
+	w.Grow(ss)
+	err = binary.Write(w, binary.LittleEndian, &l.env)
+	if err != nil {
+		return fmt.Errorf("error writting lk environment to buffer for validation %v", err)
+	}
+	crc := crc32.ChecksumIEEE(w.Bytes()[:ss-4]) // size of crc32 itself at the end of the structure
+	if crc != l.env.Crc32 {
+		return fmt.Errorf("lkenv:load: broken crc32 [0x%X]vs[0x%X]\n", crc, l.env.Crc32)
+	}
+	logger.Debugf("Load: validated crc32 (0x%X)", l.env.Crc32)
+	return nil
+}
+
+func (l *Env) Save() error {
+	logger.Debugf("Save")
+	w := bytes.NewBuffer(nil)
+	ss := binary.Size(l.env)
+	w.Grow(ss)
+	err := binary.Write(w, binary.LittleEndian, &l.env)
```

Comment:
Given that we write the thing once, calcualte the checksum, and then write it again with updated checksum, perhaps it would make sense to split the env struct into:
```
type SnapBootSelectImage struct {
    SnapBootSelect // sans Crc32 field
    /* crc32 value for structure */
    Crc32 uint32
}
```
Provided binary.Write/Read can handle embedded struct correctly.

Reply:
I updated implementation to avoid double write.
Reason this is kept as simple structure is to make it simple to read it in c on bootloader side.
I have played with idea to double wrap it win extra structure as you suggested but that would still mean two writes, first for inner structures to calculate crc and then write of wrapping structure. Now I actually rewind buffer by 4 bytes and re-write crc only.